lib/BoundaryValueDiffEqShooting/src/multiple_shooting.jl

```diff
@@ -92,33 +93,34 @@ function __solve_nlproblem!(
         nodes, cur_nshoot::Int, M::Int, N::Int, resida_len::Int, residb_len::Int,
         solve_internal_odes!::S, bca::B1, bcb::B2, prob, u0, ode_cache_loss_fn,
         ensemblealg, internal_ode_kwargs; kwargs...) where {B1, B2, S}
-    if __any_sparse_ad(alg.jac_alg)
-        J_proto = __generate_sparse_jacobian_prototype(
-            alg, prob.problem_type, bcresid_prototype, u0, N, cur_nshoot)
-    end
-
     resid_prototype = vcat(
         bcresid_prototype[1], similar(u_at_nodes, cur_nshoot * N), bcresid_prototype[2])
 
     loss_fn = @closure (du, u, p) -> __multiple_shooting_2point_loss!(
         du, u, p, cur_nshoot, nodes, prob, solve_internal_odes!,
         resida_len, residb_len, N, bca, bcb, ode_cache_loss_fn)
 
-    sd_bvp = alg.jac_alg.diffmode isa AutoSparse ? __sparsity_detection_alg(J_proto) :
-             NoSparsityDetection()
+    diffmode = if alg.jac_alg.diffmode isa AutoSparse
+        sparse_jacobian_prototype = __generate_sparse_jacobian_prototype(
+            alg, prob.problem_type, bcresid_prototype, u0, N, cur_nshoot)
+        AutoSparse(get_dense_ad(alg.jac_alg.diffmode),
+            sparsity_detector = ADTypes.KnownJacobianSparsityDetector(sparse_jacobian_prototype),
+            coloring_algorithm = alg.jac_alg.diffmode.coloring_algorithm)
+    else
+        alg.jac_alg.diffmode
+    end
 
     resid_prototype_cached = similar(resid_prototype)
-    jac_cache = sparse_jacobian_cache(
-        alg.jac_alg.diffmode, sd_bvp, nothing, resid_prototype_cached, u_at_nodes)
-    jac_prototype = init_jacobian(jac_cache)
+    jac_cache = DI.prepare_jacobian(nothing, resid_prototype_cached, diffmode, u_at_nodes)
```

Comment:
This will error in general. What is the rationale?

Reply:
> It is working fine because you're using it with backends that don't actually call the function during preparation

Yeah indeed, the current structure of BoundaryValueDiffEq.jl is not generalized enough to support backends like Enzyme, and to support other backends requires additional refactorizations, which exceed what this PR is expected, so let's focus on the ForwardDiff and FiniteDiff for now, I will fix other AD backend supports in following PRs

Reply:
Sounds good, I just wanted to let you know that this wasn't a supported use in general, even though it happens to work here

Reply:
Why is this `nothing`? You know the types?